comp/logs/agent/config/endpoints.go

```diff
@@ -56,9 +58,142 @@ type Endpoint struct {
 	Origin    IntakeOrigin
 }
 
-// GetUseSSL returns the UseSSL config setting
-func (e *Endpoint) GetUseSSL() bool {
-	return e.UseSSL == nil || *e.UseSSL
+// unmarshalEndpoint is used to load additional endpoints from the configuration which stored as JSON/mapstructure.
+// A different type is used than Endpoint since we want some fields to be private in Endpoint (APIKey, IsReliable, ...).
+type unmarshalEndpoint struct {
+	APIKey     string `mapstructure:"api_key" json:"api_key"`
+	IsReliable *bool  `mapstructure:"is_reliable" json:"is_reliable"`
+	UseSSL     *bool  `mapstructure:"use_ssl" json:"use_ssl"`
+
+	Endpoint `mapstructure:",squash"`
+}
+
+// NewEndpoint returns a new Endpoint with the minimal field initialized.
+func NewEndpoint(apiKey string, host string, port int, useSSL bool) Endpoint {
+	apiKey = pkgconfigutils.SanitizeAPIKey(apiKey)
+	return Endpoint{
+		apiKeyGetter: func() string { return apiKey },
+		Host:         host,
+		Port:         port,
+		useSSL:       useSSL,
+		isReliable:   true, // by default endpoints are reliable
+	}
+}
+
+// NewTCPEndpoint returns a new TCP  Endpoint based on LogsConfigKeys
+func NewTCPEndpoint(logsConfig *LogsConfigKeys) Endpoint {
```

Comment:
I don't understand the naming difference between this using "TCPEndpoint" and the next function using "HTTPEndpoint". The only thing that stands out is that this function uses `ProxyAddress` and the other has `Backoff` and `Recovery` state. Could you clarify the tcp vs http distinction?

Reply:
I'll try. They use different configuration fields. This code was moved from config to endpoints to ensure a constructor is used.